File: controlmesh/execution_grants.py
"""Tool grant snapshots: issued authorization carried with execution identity.

A :class:`ToolGrantSnapshot` is the persisted, minimal-privilege authorization for
one task.  Trusted Python ingress issues it; provider adapters map it onto their
natively enforceable surface before any provider process is created, or fail
closed with :class:`ToolGrantDenied` when the restriction cannot be enforced.

Grants are authorization state, distinct from :class:`~controlmesh.bus.envelope.
ExecutionContext` provenance: provenance says where a task came from, a grant
says what it may do.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from controlmesh.errors import CLIError

TOOL_GRANT_SCHEMA_VERSION = "controlmesh.tool_grant.v1"

_NETWORK_POLICIES = frozenset({"sandbox_default", "no_network"})
_CONFIRMATION_POLICIES = frozenset({"provider_runtime", "controller_required"})
_TOOL_TOKEN_RE = re.compile(r"^[A-Za-z0-9_][A-Za-z0-9_:\.\-]{0,63}$")
_BOUNDED_TOKEN_RE = re.compile(r"^[^\r\n\x00]{0,128}$")
_ROOT_TOKEN_RE = re.compile(r"^[A-Za-z0-9_\.\-/]{1,256}$")

_GRANT_SURFACES = frozenset(
    {
        "",
        "claude_tool_flags",
        "codex_sandbox",
        "gemini_policy",
        "opencode_config",
    }
)


@dataclass(frozen=True, slots=True)
class ToolGrantSnapshot:
    """Minimal-privilege authorization bound to one task and its episodes.

    ``provider_surface`` records the surface the issuing ingress validated for
    the initial execution.  It is metadata, never authorization: an adapter
    always re-derives whether it can honor the restrictions on its own surface.
    """

    tool_allow: tuple[str, ...] = ()
    tool_deny: tuple[str, ...] = ()
    network_policy: str = "sandbox_default"
    writable_roots: tuple[str, ...] = ()
    confirmation_policy: str = "provider_runtime"
    provider_surface: str = ""
    reply_transport: str = ""
    reply_chat: str = ""
    reply_topic: str = ""
    reply_thread: str = ""

    @property
    def restrictive(self) -> bool:
        """Whether the grant carries a restriction an adapter must enforce."""
        return bool(
            self.tool_allow
            or self.tool_deny
            or self.writable_roots
            or self.network_policy == "no_network"
        )

    def to_dict(self) -> dict[str, object]:
        return {
            "schema_version": TOOL_GRANT_SCHEMA_VERSION,
            "tool_allow": list(self.tool_allow),
            "tool_deny": list(self.tool_deny),
            "network_policy": self.network_policy,
            "writable_roots": list(self.writable_roots),
            "confirmation_policy": self.confirmation_policy,
            "provider_surface": self.provider_surface,
            "reply_transport": self.reply_transport,
            "reply_chat": self.reply_chat,
            "reply_topic": self.reply_topic,
            "reply_thread": self.reply_thread,
        }

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> ToolGrantSnapshot:
        if not isinstance(raw, dict):
            raise ValueError("tool grant must be an object")
        if raw.get("schema_version") != TOOL_GRANT_SCHEMA_VERSION:
            raise ValueError("unsupported persisted tool grant schema_version")
        return cls(
            tool_allow=_tool_tuple(raw.get("tool_allow")),
            tool_deny=_tool_tuple(raw.get("tool_deny")),
            network_policy=_bounded_enum(raw.get("network_policy"), _NETWORK_POLICIES, "network_policy"),
            writable_roots=_root_tuple(raw.get("writable_roots")),
            confirmation_policy=_bounded_enum(
                raw.get("confirmation_policy"), _CONFIRMATION_POLICIES, "confirmation_policy"
            ),
            provider_surface=_bounded_token(raw.get("provider_surface", ""), "provider_surface"),
            reply_transport=_bounded_token(raw.get("reply_transport", ""), "reply_transport"),
            reply_chat=_bounded_token(raw.get("reply_chat", ""), "reply_chat"),
            reply_topic=_bounded_token(raw.get("reply_topic", ""), "reply_topic"),
            reply_thread=_bounded_token(raw.get("reply_thread", ""), "reply_thread"),
        )


def _tool_tuple(raw: object) -> tuple[str, ...]:
    if raw is None:
        return ()
    if not isinstance(raw, (list, tuple)):
        raise ValueError("tool grant tool list must be an array")
    items: list[str] = []
    for item in raw:
        token = str(item)
        if not _TOOL_TOKEN_RE.fullmatch(token):
            raise ValueError(f"invalid persisted tool grant tool token: {token!r}")
        if token not in items:
            items.append(token)
    return tuple(items)


def _root_tuple(raw: object) -> tuple[str, ...]:
    if raw is None:
        return ()
    if not isinstance(raw, (list, tuple)):
        raise ValueError("tool grant writable_roots must be an array")
    items: list[str] = []
    for item in raw:
        token = str(item)
        if not _ROOT_TOKEN_RE.fullmatch(token) or ".." in token.split("/"):
            raise ValueError(f"invalid persisted tool grant writable root: {token!r}")
        if token not in items:
            items.append(token)
    return tuple(items)


def _bounded_token(raw: object, name: str) -> str:
    token = str(raw or "")
    if not _BOUNDED_TOKEN_RE.fullmatch(token):
        raise ValueError(f"invalid persisted tool grant {name}")
    return token


def _bounded_enum(raw: object, allowed: frozenset[str], name: str) -> str:
    token = str(raw or "")
    if token not in allowed:
        raise ValueError(f"invalid persisted tool grant {name}: {token!r}")
    return token


@dataclass(frozen=True, slots=True)
class ToolGrantMapping:
    """The natively enforceable translation of one grant for one provider."""

    provider: str
    surface: str
    flags: tuple[str, ...] = ()


class ToolGrantDenied(CLIError):  # noqa: N818
    """Raised before provider construction when a grant cannot be enforced."""

    def __init__(self, provider: str, reason_code: str) -> None:
        self.provider = provider
        self.reason_code = reason_code
        super().__init__(f"tool_grant_denied:{provider}:{reason_code}")

    @property
    def user_message(self) -> str:
        return (
            "This task carries a tool grant that this provider cannot enforce "
            "natively, so no provider process was started. "
            f"provider={self.provider} reason={self.reason_code}"
        )


def issue_tool_grant(
    *,
    tool_allow: tuple[str, ...] | list[str] = (),
    tool_deny: tuple[str, ...] | list[str] = (),
    network_policy: str = "sandbox_default",
    writable_roots: tuple[str, ...] | list[str] = (),
    confirmation_policy: str = "provider_runtime",
    provider_surface: str = "",
    reply_transport: str = "",
    reply_chat: object = "",
    reply_topic: object = "",
    reply_thread: object = "",
) -> ToolGrantSnapshot:
    """Issue a grant from trusted ingress, validating and normalizing input."""
    snapshot = ToolGrantSnapshot.from_dict(
        {
            "schema_version": TOOL_GRANT_SCHEMA_VERSION,
            "tool_allow": list(tool_allow),
            "tool_deny": list(tool_deny),
            "network_policy": network_policy,
            "writable_roots": list(writable_roots),
            "confirmation_policy": confirmation_policy,
            "provider_surface": provider_surface,
            "reply_transport": reply_transport,
            "reply_chat": str(reply_chat or ""),
            "reply_topic": str(reply_topic or ""),
            "reply_thread": str(reply_thread or ""),
        }
    )
    if snapshot.provider_surface and snapshot.provider_surface not in _GRANT_SURFACES:
        raise ValueError(f"unknown tool grant provider_surface: {snapshot.provider_surface!r}")
    return snapshot


@dataclass(frozen=True, slots=True)
class _GrantInput:
    provider: str
    grant: ToolGrantSnapshot
    config_allowed: tuple[str, ...]
    config_disallowed: tuple[str, ...]
    config_permission_mode: str
    config_sandbox_mode: str
    config_cli_parameters: tuple[str, ...] = ()


def map_tool_grant(
    provider: str,
    grant: ToolGrantSnapshot | None,
    *,
    config_allowed: tuple[str, ...] | list[str] = (),
    config_disallowed: tuple[str, ...] | list[str] = (),
    config_permission_mode: str = "",
    config_sandbox_mode: str = "",
    config_cli_parameters: tuple[str, ...] | list[str] = (),
) -> ToolGrantMapping:
    """Translate a grant onto *provider*'s natively enforceable surface.

    Raises :class:`ToolGrantDenied` before any provider process is created when
    the restriction cannot be enforced natively.  A grant never widens static
    configuration: denies are unioned, allows are intersected, and a grant
    combined with bypass/full-access configuration is rejected.
    """
    normalized = _GrantInput(
        provider=provider,
        grant=grant if grant is not None else _FLOOR_GRANT,
        config_allowed=tuple(config_allowed),
        config_disallowed=tuple(config_disallowed),
        config_permission_mode=str(config_permission_mode or ""),
        config_sandbox_mode=str(config_sandbox_mode or ""),
        config_cli_parameters=tuple(config_cli_parameters),
    )
    if grant is None or not grant.restrictive:
        return ToolGrantMapping(provider=provider, surface=f"{provider}_floor")
    if grant.confirmation_policy == "controller_required":
        # A restrictive task claiming controller approval requires real
        # approval evidence, which no reusable mechanism provides yet.
        raise ToolGrantDenied(provider, "controller_approval_unavailable")
    handler = _MAPPERS.get(provider)
    if handler is None:
        raise ToolGrantDenied(provider, "surface_unproven")
    _reject_override_conflicts(normalized)
    return handler(normalized)


def _reject_override_conflicts(data: _GrantInput) -> None:
    """Reject CLI parameter overrides that would defeat the grant's limits."""
    lowered = tuple(item.lower() for item in data.config_cli_parameters)
    if data.provider == "codex":
        joined = " ".join(lowered)
        if "sandbox" in joined or "network_access" in joined:
            raise ToolGrantDenied("codex", "override_conflicts_grant")
    if data.provider == "claude":
        for item in lowered:
            if item.startswith("--dangerously") or item.startswith("--permission-mode"):
                raise ToolGrantDenied("claude", "override_conflicts_grant")


_FLOOR_GRANT = ToolGrantSnapshot()


def _map_claude(data: _GrantInput) -> ToolGrantMapping:
    grant = data.grant
    if data.config_permission_mode == "bypassPermissions":
        raise ToolGrantDenied("claude", "bypass_conflicts_grant")
    if grant.network_policy == "no_network":
        # Claude flags cannot isolate workload network access; denying two
        # built-in tools is not network isolation (R3).
        raise ToolGrantDenied("claude", "no_network_unenforceable")
    if grant.tool_allow:
        raise ToolGrantDenied("claude", "allowlist_not_enforceable_flags")
    if grant.writable_roots:
        raise ToolGrantDenied("claude", "writable_roots_needs_container_mounts")
    denies: list[str] = []
    for token in (*data.config_disallowed, *grant.tool_deny):
        if token not in denies:
            denies.append(token)
    flags: list[str] = []
    if denies:
        flags += ["--disallowedTools", *denies]
    return ToolGrantMapping(provider="claude", surface="claude_tool_flags", flags=tuple(flags))


def _map_codex(data: _GrantInput) -> ToolGrantMapping:
    grant = data.grant
    if grant.tool_allow or grant.tool_deny:
        raise ToolGrantDenied("codex", "tool_granularity_unsupported")
    if grant.writable_roots:
        raise ToolGrantDenied("codex", "writable_roots_unverified")
    if data.config_permission_mode == "bypassPermissions":
        raise ToolGrantDenied("codex", "bypass_conflicts_grant")
    if grant.network_policy == "no_network":
        if data.config_sandbox_mode == "full-access":
            raise ToolGrantDenied("codex", "no_network_conflicts_full_access")
        if data.config_sandbox_mode == "workspace-write":
            return ToolGrantMapping(
                provider="codex",
                surface="codex_sandbox",
                flags=("-c", 'sandbox_workspace_write={"network_access": false}'),
            )
        if data.config_sandbox_mode != "read-only":
            raise ToolGrantDenied("codex", "no_network_unproven_sandbox")
    return ToolGrantMapping(provider="codex", surface="codex_sandbox")


def _map_gemini(data: _GrantInput) -> ToolGrantMapping:
    grant = data.grant
    if grant.tool_allow or grant.tool_deny:
        raise ToolGrantDenied("gemini", "policy_engine_config_unverified")
    if grant.network_policy == "no_network" or grant.writable_roots:
        raise ToolGrantDenied("gemini", "policy_engine_config_unverified")
    return ToolGrantMapping(provider="gemini", surface="gemini_policy")


def _map_opencode(_data: _GrantInput) -> ToolGrantMapping:
    raise ToolGrantDenied("opencode", "config_overlay_unverified")


_MAPPERS = {
    "claude": _map_claude,
    "codex": _map_codex,
    "gemini": _map_gemini,
    "opencode": _map_opencode,
}


class ReplyTargetMismatch(CLIError):  # noqa: N818
    """Raised before delivery when the target diverges from the granted one."""

    def __init__(self, field: str) -> None:
        self.field = field
        super().__init__(f"reply_target_mismatch:{field}")

    @property
    def user_message(self) -> str:
        return (
            "Result delivery was blocked because the delivery target does not "
            "match the task's trusted reply identity. "
            f"field={self.field}"
        )


def validate_reply_target(
    grant: ToolGrantSnapshot | None,
    *,
    transport: str,
    chat_id: object,
    topic_id: object = "",
    thread_id: object = "",
) -> None:
    """Verify delivery identity matches the grant's trusted reply identity.

    Only fields the grant actually pins are checked; a grant without reply
    identity (legacy records) accepts the trusted delivery target unchanged.
    """
    if grant is None:
        return
    checks = (
        ("reply_transport", transport),
        ("reply_chat", str(chat_id or "")),
        ("reply_topic", str(topic_id or "")),
        ("reply_thread", str(thread_id or "")),
    )
    for field_name, granted_value in checks:
        expected = getattr(grant, field_name)
        if expected and str(granted_value or "") != expected:
            raise ReplyTargetMismatch(field_name)


def issue_task_grant_for_submit(
    *,
    source_scope: str,
    requested_tool_deny: tuple[str, ...] | list[str] = (),
    requested_no_network: bool = False,
    transport: str,
    chat_id: object = "",
    topic_id: object = "",
    thread_id: object = "",
) -> ToolGrantSnapshot:
    """Issue the submit-time grant from the source floor plus narrowing requests.

    Trusted ingress only: restriction requests may only narrow (deny list and
    network isolation); there is no way to request widened permissions through
    this path, and message content never reaches it.
    """
    from controlmesh.execution_policy import SOURCE_SCOPES_REQUIRING_SANDBOX, SourceScope

    try:
        sandbox_required = SourceScope(str(source_scope)) in SOURCE_SCOPES_REQUIRING_SANDBOX
    except ValueError:
        sandbox_required = False
    floor_confirmation = "controller_required" if sandbox_required else "provider_runtime"
    return issue_tool_grant(
        tool_deny=tuple(requested_tool_deny),
        network_policy="no_network" if requested_no_network else "sandbox_default",
        confirmation_policy=floor_confirmation,
        reply_transport=str(transport or ""),
        reply_chat=str(chat_id or ""),
        reply_topic=str(topic_id or ""),
        reply_thread=str(thread_id or ""),
    )

File: packages/controlmesh-runtime-core/src/execution-grants.ts
import { decodeSourceScope, type SourceScope } from "./execution-context";
import { sourceRequiresSandbox, type ExecutionPolicyDecision } from "./execution-policy";
import { object, requireThat } from "./value";

export interface ToolGrantSnapshot {
  readonly schema_version: "controlmesh.tool_grant.v1";
  readonly tool_allow: readonly string[];
  readonly tool_deny: readonly string[];
  readonly network_policy: "sandbox_default" | "no_network";
  readonly writable_roots: readonly string[];
  readonly confirmation_policy: "provider_runtime" | "controller_required";
  readonly provider_surface: string;
  readonly reply_transport: string;
  readonly reply_chat: string;
  readonly reply_topic: string;
  readonly reply_thread: string;
}
type GrantInput = Partial<Omit<ToolGrantSnapshot, "schema_version">>;
const surfaces = ["", "claude_tool_flags", "codex_sandbox", "gemini_policy", "opencode_config"];
function token(value: unknown): string {
  const text = value ?? "";
  requireThat(typeof text === "string" && /^[^\r\n\x00]{0,128}$/u.test(text), "invalid_tool_grant_identity");
  return text;
}
function tokens(value: unknown, roots = false): readonly string[] {
  if (value === undefined || value === null) return Object.freeze([]);
  requireThat(Array.isArray(value) && value.length <= 256, "invalid_tool_grant_list");
  const pattern = roots ? /^[A-Za-z0-9_.\-/]{1,256}$/ : /^[A-Za-z0-9_][A-Za-z0-9_:.\-]{0,63}$/;
  requireThat(value.every(item => typeof item === "string" && pattern.test(item) && (!roots || !item.split("/").includes(".."))), "invalid_tool_grant_token");
  return Object.freeze([...new Set(value as string[])]);
}
function configuredDenials(value: readonly string[] | undefined): readonly string[] {
  const list = value ?? [];
  // Native static config supports expressions such as Bash(rm *), unlike the portable grant token format.
  requireThat(Array.isArray(list) && list.length <= 256 && list.every(item => typeof item === "string"
    && item.length > 0 && item.length <= 1024 && !/^[\-]|[\r\n\x00]/.test(item)), "invalid_configured_tool_denial");
  return list;
}

/** Shape compatibility for persisted v1, with strict types and bounded lists at the new TS boundary. */
export function decodeToolGrant(raw: unknown): ToolGrantSnapshot {
  requireThat(object(raw) && raw.schema_version === "controlmesh.tool_grant.v1", "issued_tool_grant_required");
  requireThat(raw.network_policy === "sandbox_default" || raw.network_policy === "no_network", "invalid_tool_grant_network");
  requireThat(raw.confirmation_policy === "provider_runtime" || raw.confirmation_policy === "controller_required", "invalid_tool_grant_confirmation");
  return Object.freeze({ schema_version: "controlmesh.tool_grant.v1", tool_allow: tokens(raw.tool_allow), tool_deny: tokens(raw.tool_deny), writable_roots: tokens(raw.writable_roots, true),
    network_policy: raw.network_policy, confirmation_policy: raw.confirmation_policy, provider_surface: token(raw.provider_surface),
    reply_transport: token(raw.reply_transport), reply_chat: token(raw.reply_chat), reply_topic: token(raw.reply_topic), reply_thread: token(raw.reply_thread) });
}

/** Trusted local issuance only. A provider_surface label never proves enforcement. */
export function issueToolGrant(input: GrantInput = {}): ToolGrantSnapshot {
  const grant = decodeToolGrant({ network_policy: "sandbox_default", confirmation_policy: "provider_runtime", ...input, schema_version: "controlmesh.tool_grant.v1" });
  requireThat(surfaces.includes(grant.provider_surface), "unknown_tool_grant_surface");
  return grant;
}

export function restrictiveGrant(grant: ToolGrantSnapshot): boolean {
  return !!(grant.tool_allow.length || grant.tool_deny.length || grant.writable_roots.length || grant.network_policy === "no_network");
}

export interface ToolGrantMapping { provider: string; surface: string; flags: readonly string[] }
export interface ProviderGrantConfig {
  config_allowed?: readonly string[];
  config_disallowed?: readonly string[];
  config_permission_mode?: string;
  config_sandbox_mode?: string;
  config_cli_parameters?: readonly string[];
}
export class ToolGrantDenied extends Error {
  constructor(public readonly provider: string, public readonly reason_code: string) {
    super(`tool_grant_denied:${provider}:${reason_code}`); this.name = "ToolGrantDenied";
  }
}

/** Native flag mapping, matching the existing Python golden. This is NOT a process admission decision. */
export function mapToolGrant(provider: string, raw: ToolGrantSnapshot | null, config: ProviderGrantConfig = {}): ToolGrantMapping {
  requireThat(typeof provider === "string" && /^[a-z0-9_\-]{1,64}$/.test(provider), "invalid_grant_provider");
  const grant = raw === null ? null : decodeToolGrant(raw);
  const mapping = (surface: string, flags: readonly string[] = []): ToolGrantMapping => Object.freeze({ provider, surface, flags: Object.freeze([...flags]) });
  // Historical semantics retained for differential checks; admission below separately checks confirmation.
  if (grant === null || !restrictiveGrant(grant)) return mapping(`${provider}_floor`);
  const deny = (reason: string): never => { throw new ToolGrantDenied(provider, reason); };
  if (grant.confirmation_policy === "controller_required") deny("controller_approval_unavailable");
  if (!["claude", "codex", "gemini", "opencode"].includes(provider)) deny("surface_unproven");
  const parameters = config.config_cli_parameters ?? [];
  requireThat(Array.isArray(parameters) && parameters.length <= 256 && parameters.every(p => typeof p === "string" && p.length <= 4096 && !/[\x00\r\n]/.test(p)), "invalid_provider_parameters");
  const lowered = parameters.map(p => p.toLowerCase());
  if (provider === "codex" && /sandbox|network_access/.test(lowered.join(" "))) deny("override_conflicts_grant");
  if (provider === "claude" && lowered.some(p => p.startsWith("--dangerously") || p.startsWith("--permission-mode"))) deny("override_conflicts_grant");
  if (provider === "claude") {
    if (config.config_permission_mode === "bypassPermissions") deny("bypass_conflicts_grant");
    if (grant.network_policy === "no_network") deny("no_network_unenforceable");
    if (grant.tool_allow.length) deny("allowlist_not_enforceable_flags");
    if (grant.writable_roots.length) deny("writable_roots_needs_container_mounts");
    const denies = [...new Set([...configuredDenials(config.config_disallowed), ...grant.tool_deny])];
    return mapping("claude_tool_flags", denies.length ? ["--disallowedTools", ...denies] : []);
  }
  if (provider === "codex") {
    if (grant.tool_allow.length || grant.tool_deny.length) deny("tool_granularity_unsupported");
    if (grant.writable_roots.length) deny("writable_roots_unverified");
    if (config.config_permission_mode === "bypassPermissions") deny("bypass_conflicts_grant");
    if (grant.network_policy === "no_network") {
      if (config.config_sandbox_mode === "full-access") deny("no_network_conflicts_full_access");
      if (config.config_sandbox_mode === "workspace-write") return mapping("codex_sandbox", ["-c", 'sandbox_workspace_write={"network_access": false}']);
      if (config.config_sandbox_mode !== "read-only") deny("no_network_unproven_sandbox");
    }
    return mapping("codex_sandbox");
  }
  if (provider === "gemini") deny("policy_engine_config_unverified");
  return deny("config_overlay_unverified");
}

/** Admission is stricter than flag mapping: even an otherwise empty grant cannot waive a source/approval floor. */
export function enforceProviderConfirmation(provider: string, grant: ToolGrantSnapshot, policy?: ExecutionPolicyDecision): void {
  const current = decodeToolGrant(grant);
  requireThat(policy === undefined || policy.outcome === "accepted", "execution_policy_not_accepted");
  if (current.confirmation_policy === "controller_required" || policy?.confirmation_policy === "controller_required") {
    throw new ToolGrantDenied(provider, "controller_approval_unavailable");
  }
}

export interface SubmitGrantInput {
  source_scope: SourceScope;
  requested_tool_deny?: readonly string[];
  requested_no_network?: boolean;
  transport: string;
  chat_id?: string;
  topic_id?: string;
  thread_id?: string;
}
/** Source-derived floor plus narrowing requests only. Unknown ingress never falls back to host compatibility. */
export function issueTaskGrantForSubmit(input: SubmitGrantInput): ToolGrantSnapshot {
  const scope = decodeSourceScope(input.source_scope);
  requireThat(input.requested_no_network === undefined || typeof input.requested_no_network === "boolean", "invalid_requested_network_policy");
  return issueToolGrant({ tool_deny: input.requested_tool_deny ?? [], network_policy: input.requested_no_network ? "no_network" : "sandbox_default",
    confirmation_policy: sourceRequiresSandbox(scope) ? "controller_required" : "provider_runtime", reply_transport: input.transport,
    reply_chat: input.chat_id ?? "", reply_topic: input.topic_id ?? "", reply_thread: input.thread_id ?? "" });
}

export class ReplyTargetMismatch extends Error {
  constructor(public readonly field: string) { super(`reply_target_mismatch:${field}`); this.name = "ReplyTargetMismatch"; }
}
export function validateReplyTarget(raw: ToolGrantSnapshot | null, target: { transport: string; chat_id: string; topic_id?: string; thread_id?: string }): void {
  if (raw === null) return;
  const grant = decodeToolGrant(raw);
  const checks = { reply_transport: target.transport, reply_chat: target.chat_id, reply_topic: target.topic_id ?? "", reply_thread: target.thread_id ?? "" };
  for (const [field, value] of Object.entries(checks)) {
    requireThat(typeof value === "string", "invalid_reply_target");
    const expected = grant[field as keyof typeof checks];
    if (expected && value !== expected) throw new ReplyTargetMismatch(field);
  }
}
